nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ExtractRecordSchema.java

```diff
@@ -0,0 +1,134 @@
+public class ExtractRecordSchema extends AbstractProcessor {
+
+    public static final String SCHEMA_ATTRIBUTE_NAME = "avro.schema";
+
+    static final PropertyDescriptor RECORD_READER = new PropertyDescriptor.Builder()
+            .name("record-reader")
+            .displayName("Record Reader")
+            .description("Specifies the Controller Service to use for reading incoming data")
+            .identifiesControllerService(RecordReaderFactory.class)
+            .required(true)
+            .build();
+
+    static final Relationship REL_SUCCESS = new Relationship.Builder()
+            .name("success")
+            .description("FlowFiles whose record schemas are successfully extracted will be routed to this relationship")
+            .build();
+    static final Relationship REL_FAILURE = new Relationship.Builder()
+            .name("failure")
+            .description("If a FlowFile's record schema cannot be extracted from the configured input format, "
+                    + "the FlowFile will be routed to this relationship")
+            .build();
+
+    private final LoadingCache<RecordSchema, String> avroSchemaTextCache = Caffeine.newBuilder()
+            .maximumSize(10)
+            .build(schema -> AvroTypeUtil.extractAvroSchema(schema).toString());
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        final List<PropertyDescriptor> properties = new ArrayList<>();
+        properties.add(RECORD_READER);
+        return properties;
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        final Set<Relationship> relationships = new HashSet<>();
+        relationships.add(REL_SUCCESS);
+        relationships.add(REL_FAILURE);
+        return relationships;
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        final RecordReaderFactory readerFactory = context.getProperty(RECORD_READER).asControllerService(RecordReaderFactory.class);
+        final Map<String, String> originalAttributes = flowFile.getAttributes();
+        final RecordSchema recordSchema;
+        try (final InputStream inputStream = session.read(flowFile);
+             final RecordReader reader = readerFactory.createRecordReader(originalAttributes, inputStream, flowFile.getSize(), getLogger())) {
+            recordSchema = reader.getSchema();
+
+        } catch (final Exception e) {
+            getLogger().error("Failed to process {}; will route to failure", flowFile, e);
+            // Since we are wrapping the exceptions above there should always be a cause
+            // but it's possible it might not have a message. This handles that by logging
+            // the name of the class thrown.
+            Throwable c = e.getCause();
+            if (c != null) {
```

Comment:
Recommend reverse the blocks so the that conditional reads `if (cause == null) { ... } else { ... }`